File: geocode.js
/**
 * Geocode a Vietnamese address to latitude/longitude coordinates
 * Uses Nominatim (OpenStreetMap) - FREE, no API key required
 *
 * Usage: node geocode.js "161D/104/23G lạc long quân, phường bình thới, hồ chí minh"
 */

const https = require('https');

function geocodeAddress(address) {
  return new Promise((resolve, reject) => {
    // Build the Nominatim API URL
    const encodedAddress = encodeURIComponent(address);
    const url = `https://nominatim.openstreetmap.org/search?q=${encodedAddress}&format=json&limit=1&countrycodes=vn`;

    const options = {
      headers: {
        'User-Agent': 'Polomanor-Geocoder/1.0 (contact@example.com)'
      }
    };

    https.get(url, options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        // Debug raw response if parse fails
        if (data.trim().startsWith('Access') || data.trim().startsWith('<!')) {
          reject(new Error('Nominatim API chặn yêu cầu (có thể do rate limit). Đang thử API dự phòng...'));
          return;
        }
        try {
          const results = JSON.parse(data);
          if (results && results.length > 0) {
            const result = results[0];
            resolve({
              address: result.display_name,
              latitude: parseFloat(result.lat),
              longitude: parseFloat(result.lon),
              boundingBox: result.boundingbox
            });
          } else {
            reject(new Error('Không tìm thấy tọa độ cho địa chỉ này.'));
          }
        } catch (err) {
          console.log('Raw response:', data.substring(0, 200));
          reject(new Error('Lỗi parse phản hồi từ server: ' + err.message));
        }
      });
    }).on('error', (err) => {
      reject(new Error('Lỗi kết nối: ' + err.message));
    });
  });
}

// Main execution
async function main() {
  // Lấy địa chỉ từ command line arguments, hoặc dùng mặc định
  const inputAddress = process.argv[2] || '161D/104/23G lạc long quân, phường bình thới, hồ chí minh';

  console.log('============================================');
  console.log('  GEOCODING - Chuyển địa chỉ → Tọa độ');
  console.log('============================================\n');
  console.log('Địa chỉ cần tìm:');
  console.log('  "' + inputAddress + '"\n');
  console.log('Đang gửi yêu cầu đến Nominatim (OpenStreetMap)...\n');

  try {
    const result = await geocodeAddress(inputAddress);

    console.log('✅ TÌM THẤY TỌA ĐỘ!\n');
    console.log('---------------------------------------------');
    console.log('  Vĩ độ (Latitude) : ' + result.latitude);
    console.log('  Kinh độ (Longitude): ' + result.longitude);
    console.log('---------------------------------------------\n');
    console.log('Địa chỉ đầy đủ từ Nominatim:');
    console.log('  ' + result.address + '\n');
    console.log('Google Maps URL:');
    console.log('  https://www.google.com/maps?q=' + result.latitude + ',' + result.longitude + '\n');
    console.log('OpenStreetMap URL:');
    console.log('  https://www.openstreetmap.org/?mlat=' + result.latitude + '&mlon=' + result.longitude + '#map=18/' + result.latitude + '/' + result.longitude + '\n');

  } catch (err) {
    console.error('❌ LỖI: ' + err.message + '\n');
    process.exit(1);
  }
}

main();

File: geocode.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Chuyển địa chỉ tiếng Việt thành tọa độ (lat/lng)
Sử dụng Nominatim (OpenStreetMap) - MIỄN PHÍ, không cần API key

Cách dùng:
    python geocode.py "161D/104/23G lạc long quân, phường bình thới, hồ chí minh"
"""

import urllib.request
import urllib.parse
import json
import sys
import ssl


def geocode_address(address: str):
    """Gọi Nominatim API để lấy tọa độ từ địa chỉ."""
    encoded = urllib.parse.quote(address)
    url = f"https://nominatim.openstreetmap.org/search?format=json&q={encoded}&limit=1&countrycodes=vn"

    req = urllib.request.Request(
        url,
        headers={
            "User-Agent": "Polomanor-Geocoder/1.0 (contact@polomanor.vn)",
            "Accept-Language": "vi",
        }
    )

    # Tạo SSL context an toàn
    ctx = ssl.create_default_context()

    try:
        with urllib.request.urlopen(req, context=ctx, timeout=15) as response:
            data = json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        print(f"❌ HTTP Error {e.code}: {e.reason}")
        print(f"   Có thể API đã chặn yêu cầu. Hãy thử mở file test-geocode.html trong trình duyệt.")
        sys.exit(1)
    except Exception as e:
        print(f"❌ Lỗi kết nối: {e}")
        sys.exit(1)

    if not data:
        print("❌ Không tìm thấy tọa độ cho địa chỉ này.")
        sys.exit(1)

    result = data[0]
    return {
        "address": result["display_name"],
        "lat": float(result["lat"]),
        "lon": float(result["lon"]),
    }


def main():
    address = sys.argv[1] if len(sys.argv) > 1 else "161D/104/23G lạc long quân, phường bình thới, hồ chí minh"

    print("=" * 50)
    print("  GEOCODING - Chuyển địa chỉ → Tọa độ")
    print("=" * 50)
    print()
    print(f"Địa chỉ cần tìm:\n  \"{address}\"")
    print()
    print("Đang gửi yêu cầu đến Nominatim (OpenStreetMap)...")
    print()

    result = geocode_address(address)

    print("✅ TÌM THẤY TỌA ĐỘ!\n")
    print("-" * 46)
    print(f"  Vĩ độ  (Latitude) :  {result['lat']:.6f}")
    print(f"  Kinh độ (Longitude): {result['lon']:.6f}")
    print("-" * 46)
    print()
    print(f"Địa chỉ đầy đủ:\n  {result['address']}\n")
    print("Google Maps URL:")
    print(f"  https://www.google.com/maps?q={result['lat']},{result['lon']}\n")
    print("OpenStreetMap URL:")
    print(f"  https://www.openstreetmap.org/?mlat={result['lat']}&mlon={result['lon']}#map=18/{result['lat']}/{result['lon']}\n")


if __name__ == "__main__":
    main()

